lib/utils/common.py

```diff
@@ -235,36 +98,391 @@ def _check_value(data={}, keys=()):
+            check_func (CheckCallback): A callback that determines whether the feature is present.
+            fields (tuple[str, ...]): All the fields to check.
+            compare_op (CompareCallable, optional): A function that tells how we should compare the field value to compare_val. Defaults to operator.gt.
+            compare_val (CompareValue, optional): The value to use in the comparison to each field value. Defaults to 0.
+        """
+        self._source = source
+        self._fields: tuple[str, ...] = fields
+        self._check_func = check_func
+        self._compare_op = compare_op
+        self._compare_val = compare_val
+
+    def check(self, sources: dict[str, Any]) -> bool:
+        if self._source not in sources:
+            return False
+
+        return self._check_func(
+            sources[self._source], self._fields, self._compare_op, self._compare_val
+        )
+
+
+class GlobalFieldCheck(FieldCheck):
+    def __init__(
+        self,
+        source: str,
+        fields: tuple[str, ...],
+        compare_op: CompareCallable = operator.gt,
+        compare_val: CompareValue = 0,
+    ):
+        super().__init__(
+            source, _check_feature_by_keys, fields, compare_op, compare_val
+        )
+
+
+class NestedFieldCheck(FieldCheck):
+    def __init__(
+        self,
+        source: str,
+        fields: tuple[str, ...],
+        compare_op: CompareCallable = operator.gt,
+        compare_val: CompareValue = 0,
+    ):
+        super().__init__(
+            source, _check_nested_feature_by_keys, fields, compare_op, compare_val
+        )
+
+
+class ServiceFieldCheck(GlobalFieldCheck):
+    pass
+
+
+class SecurityFieldCheck(GlobalFieldCheck):
+    pass
+
+
+class NamespacesFieldCheck(NestedFieldCheck):
+    pass
+
+
+class XdrDcsFieldCheck(NestedFieldCheck):
+    pass
+
+
+class FeatureCheck:
+    def __init__(self, display_key: str, *field_checks: FieldCheck) -> None:
+        """A collection of checks that indicate a feature is in use.
+
+        Args:
+            display_key (str): The value to be displayed when a feature is determined to be in use.
+            field_checks (tuple[FieldCheck, ...]): All the possible checks that would indicate that this
+            feature is in use. Sometimes different metrics may indicate an feature is in use like
+            errors, timeout, success. It also might be the case that a stat or config no longer exists
+            and having multiple checks allows for backwards compatibility.
+        """
+        self.display_key = display_key
+        self._field_checks = field_checks
+
+    def check(self, sources: dict[str, Any]) -> bool:
+        if not sources:
+            return False
+
+        for field_check in self._field_checks:
+            if field_check.check(sources):
+                return True
+
+        return False
+
+    def __str__(self):
+        for field_check in self._field_checks:
+            return True
```

Comment:
Is this incomplete?

Reply:
yes, good catch. Ill remove it. I was going to automate the internal docs creation but decided against it